File: kzen-auto-jvm/src/main/java/tech/kzen/auto/server/objects/report/pipeline/input/model/RecordItemBuffer.java
package tech.kzen.auto.server.objects.report.pipeline.input.model;


import net.openhft.hashing.LongHashFunction;
import tech.kzen.auto.server.objects.report.pipeline.input.parse.CsvRecordParser;
import tech.kzen.auto.server.objects.report.pipeline.input.parse.NumberParseUtils;
import tech.kzen.auto.server.objects.report.pipeline.input.parse.TsvRecordParser;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;


public class RecordItemBuffer
{
    //-----------------------------------------------------------------------------------------------------------------
    private static final double doubleCacheMissing = -0.0;
    private static final long missingNumberBits = Double.doubleToRawLongBits(doubleCacheMissing);

    private static boolean isDoubleCacheMissing(double value) {
        return Double.doubleToRawLongBits(value) == missingNumberBits;
    }


    public static RecordItemBuffer of(String... values) {
        return of(Arrays.asList(values));
    }


    public static RecordItemBuffer of(List<String> values) {
        RecordItemBuffer buffer = new RecordItemBuffer(0, 0);
        buffer.addAll(values);
        buffer.populateCaches();
        return buffer;
    }


    public static RecordItemBuffer ofSingle(char[] contents, int offset, int length) {
        RecordItemBuffer buffer = new RecordItemBuffer(length, 1);
        System.arraycopy(contents, offset, buffer.fieldContents, 0, length);
        buffer.fieldCount = 1;
        buffer.fieldEnds[0] = length;
        buffer.fieldContentLength = length;
        buffer.nonEmpty = true;
        buffer.populateCaches();
        return buffer;
    }


    //-----------------------------------------------------------------------------------------------------------------
    char[] fieldContents;
    int[] fieldEnds;
    double[] doublesCache;
    long[] hashesCache;
    private int fieldCount = 0;
    private int fieldContentLength = 0;
    private boolean nonEmpty = false;


    //-----------------------------------------------------------------------------------------------------------------
    public RecordItemBuffer()
    {
        this(0, 0);
    }


    public RecordItemBuffer(int expectedContentLength, int expectedFieldCount)
    {
        fieldContents = new char[expectedContentLength];
        fieldEnds = new int[expectedFieldCount];
    }


    //-----------------------------------------------------------------------------------------------------------------
    public int fieldCount() {
        return fieldCount;
    }


    public int fieldContentLength() {
        return fieldContentLength;
    }


    public boolean isEmpty() {
        return ! nonEmpty &&
                fieldCount <= 1 &&
                fieldContentLength == 0;
    }


    //-----------------------------------------------------------------------------------------------------------------
    public String getString(int fieldIndex) {
        int startIndex = start(fieldIndex);
        int length = fieldEnds[fieldIndex] - startIndex;
        return new String(fieldContents, startIndex, length);
    }


    public List<String> toList() {
        List<String> list = new ArrayList<>();
        for (int i = 0; i < fieldCount; i++) {
            String item = getString(i);
            list.add(item);
        }
        return list;
    }


//    /**
//     * NB: must call parseDoubles ahead of time
//     * @param fieldIndex field index
//     * @return double value of field contents or NaN
//     */
//    public double cachedDoubleOrNan(int fieldIndex) {
//        return doublesCache[fieldIndex];
//    }


    public void populateCaches() {
        growCachesIfRequired(fieldCount);

        for (int i = 0; i < fieldCount; i++) {
            populateCache(i);
        }
    }


    private void growCachesIfRequired(int size) {
        if (doublesCache == null) {
            doublesCache = new double[fieldCount];
            Arrays.fill(doublesCache, doubleCacheMissing);

            hashesCache = new long[fieldCount];
        }
        else if (doublesCache.length < size) {
            int oldLength = doublesCache.length;
            doublesCache = Arrays.copyOf(doublesCache, size);
            Arrays.fill(doublesCache, oldLength, size, doubleCacheMissing);

            hashesCache = Arrays.copyOf(hashesCache, size);
        }
    }


    private void populateCache(int fieldIndex) {
        int start = start(fieldIndex);
        int length = fieldEnds[fieldIndex] - start;

        cacheDouble(fieldIndex, start, length);
        cacheHash(fieldIndex, start, length);
    }


    private void cacheDouble(int fieldIndex, int start, int length) {
        int startCursor = start;
        int lengthCursor = length;
        for (int i = 0; i < lengthCursor; i++) {
            if (fieldContents[startCursor] != ' ') {
                break;
            }
            startCursor++;
            lengthCursor--;
        }

        for (int i = (lengthCursor - 1); i >= 0; i--) {
            if (fieldContents[startCursor + i] != ' ') {
                break;
            }
            lengthCursor--;
        }

        double value = NumberParseUtils.toDoubleOrNan(fieldContents, startCursor, lengthCursor);
        doublesCache[fieldIndex] = value;
    }


    private void cacheHash(int fieldIndex, int start, int length) {
        long value = LongHashFunction.murmur_3().hashChars(fieldContents, start, length);
        hashesCache[fieldIndex] = value;
    }


    //-----------------------------------------------------------------------------------------------------------------
    public String toCsv() {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        try (OutputStreamWriter writer = new OutputStreamWriter(out, StandardCharsets.UTF_8)) {
            writeCsv(writer);
        }
        catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        return out.toString(StandardCharsets.UTF_8);
    }


    public void writeCsv(Writer out) throws IOException {
        if (fieldCount == 1 && fieldContentLength == 0 && nonEmpty) {
            out.write("\"\"");
            return;
        }

        for (int i = 0; i < fieldCount; i++) {
            if (i != 0) {
                out.write(CsvRecordParser.delimiterInt);
            }

            writeCsvField(i, out);
        }
    }


    public void writeCsvField(int fieldIndex, Writer out) throws IOException {
        int startIndex = start(fieldIndex);
        int endIndex = fieldEnds[fieldIndex];
        CsvRecordParser.writeCsv(fieldContents, startIndex, endIndex, out);
    }


    public String toTsv() {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        try (OutputStreamWriter writer = new OutputStreamWriter(out, StandardCharsets.UTF_8)) {
            writeTsv(writer);
        }
        catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        return out.toString(StandardCharsets.UTF_8);
    }


    private void writeTsv(Writer out) throws IOException {
        if (fieldCount == 1 && fieldContentLength == 0 && nonEmpty) {
            throw new IllegalStateException("Can't represent non-empty record with single empty column");
        }

        for (int i = 0; i < fieldCount; i++) {
            if (i != 0) {
                out.write(TsvRecordParser.delimiterInt);
            }

            writeTsvField(i, out);
        }
    }


    private void writeTsvField(int fieldIndex, Writer out) throws IOException {
        int startIndex = start(fieldIndex);
        int length = fieldEnds[fieldIndex] - startIndex;
        out.write(fieldContents, startIndex, length);
    }


    //-----------------------------------------------------------------------------------------------------------------
    public void indicateNonEmpty() {
        nonEmpty = true;
    }


    public void addToField(char nextChar) {
        growFieldContentsIfRequired(fieldContentLength + 1);
        fieldContents[fieldContentLength] = nextChar;
        fieldContentLength++;
    }


    public void addToField(char[] chars, int offset, int length) {
        if (length != 0) {
            growFieldContentsIfRequired(fieldContentLength + length);
            System.arraycopy(chars, offset, fieldContents, fieldContentLength, length);
            fieldContentLength += length;
        }
    }


    public void commitField() {
        growFieldEndsIfRequired(fieldCount);

        fieldEnds[fieldCount] = fieldContentLength;
        fieldCount++;
    }


    public void addToFieldAndCommit(char[] chars, int offset, int length) {
        if (length != 0) {
            if (fieldContents.length < fieldContentLength + length) {
                int nextSize = Math.max((int) (fieldContents.length * 1.2), fieldContentLength + length);
                fieldContents = Arrays.copyOf(fieldContents, nextSize);
            }
            System.arraycopy(chars, offset, fieldContents, fieldContentLength, length);
            fieldContentLength += length;
        }

        if (fieldEnds.length <= fieldCount) {
            int nextSize = Math.max((int) (fieldEnds.length * 1.2 + 1), fieldCount);
            fieldEnds = Arrays.copyOf(fieldEnds, nextSize);
        }
        fieldEnds[fieldCount] = fieldContentLength;
        fieldCount++;
    }


    public void add(String value) {
        growFieldContentsIfRequired(fieldContentLength + value.length());

        // see: https://stackoverflow.com/questions/8894258/fastest-way-to-iterate-over-all-the-chars-in-a-string
        for (int i = 0; i < value.length(); i++) {
            fieldContents[fieldContentLength++] = value.charAt(i);
        }

        commitField();
    }


    public void addAll(List<String> values) {
        for (String value : values) {
            add(value);
        }
    }


    public void addAllAndPopulateCaches(String[] values) {
        growCachesIfRequired(fieldCount + values.length);

        for (String value : values) {
            add(value);
            populateCache(fieldCount - 1);
        }
    }


    public void clear() {
        fieldCount = 0;
        fieldContentLength = 0;
        nonEmpty = false;
    }


    //-----------------------------------------------------------------------------------------------------------------
    public char[] fieldContentsUnsafe() {
        return fieldContents;
    }


    public int[] fieldEndsUnsafe() {
        return fieldEnds;
    }


    public void setCountAndLengthUnsafe(int fieldCount, int fieldContentLength) {
        this.fieldCount = fieldCount;
        this.fieldContentLength = fieldContentLength;
    }


    public void addToFieldUnsafe(char nextChar) {
        fieldContents[fieldContentLength] = nextChar;
        fieldContentLength++;
    }


    public void addToFieldUnsafe(char[] chars, int offset, int length) {
        System.arraycopy(chars, offset, fieldContents, fieldContentLength, length);
        fieldContentLength += length;
    }


    public void commitFieldUnsafe() {
        fieldEnds[fieldCount] = fieldContentLength;
        fieldCount++;
    }


    public void addToFieldAndCommitUnsafe(char[] chars, int offset, int length) {
        System.arraycopy(chars, offset, fieldContents, fieldContentLength, length);
        fieldContentLength += length;

        fieldEnds[fieldCount] = fieldContentLength;
        fieldCount++;
    }


    //-----------------------------------------------------------------------------------------------------------------
    public void copy(RecordItemBuffer that) {
        fieldCount = that.fieldCount;
        fieldContentLength = that.fieldContentLength;
        nonEmpty = that.nonEmpty;

        growFieldContentsIfRequired(fieldContentLength);
        growFieldEndsIfRequired(fieldCount);

        System.arraycopy(that.fieldContents, 0, fieldContents, 0, fieldContentLength);
        System.arraycopy(that.fieldEnds, 0, fieldEnds, 0, fieldCount);

        if (that.doublesCache != null) {
            if (doublesCache == null || doublesCache.length < fieldCount) {
                doublesCache = new double[fieldCount];
                hashesCache = new long[fieldCount];
            }
            System.arraycopy(that.doublesCache, 0, doublesCache, 0, fieldCount);
            System.arraycopy(that.hashesCache, 0, hashesCache, 0, fieldCount);
        }
    }


    public RecordItemBuffer prototype() {
        RecordItemBuffer prototype = new RecordItemBuffer(0, 0);
        prototype.copy(this);
        return prototype;
    }


    //-----------------------------------------------------------------------------------------------------------------
    public void growTo(int requiredLength, int requiredFieldCount) {
        growFieldContentsIfRequired(requiredLength);
        growFieldEndsIfRequired(requiredFieldCount);
    }


    public void growBy(int additionalLength, int additionalFieldCount) {
        growFieldContentsIfRequired(fieldContentLength + additionalLength);
        growFieldEndsIfRequired(fieldCount + additionalFieldCount);
    }


    private void growFieldContentsIfRequired(int required) {
        if (fieldContents.length < required) {
            int nextSize = Math.max((int) (fieldContents.length * 1.2), required);
            fieldContents = Arrays.copyOf(fieldContents, nextSize);
        }
    }


    private void growFieldEndsIfRequired(int required) {
        if (fieldEnds.length <= required) {
            int nextSize = Math.max((int) (fieldEnds.length * 1.2 + 1), required);
            fieldEnds = Arrays.copyOf(fieldEnds, nextSize);
        }
    }


    int start(int fieldIndex) {
        return fieldIndex == 0 ? 0 : fieldEnds[fieldIndex - 1];
    }


    //-----------------------------------------------------------------------------------------------------------------
    @Override
    public String toString() {
        return toCsv();
    }
}

File: kzen-auto-jvm/src/main/kotlin/tech/kzen/auto/server/objects/report/pipeline/input/model/RecordItemBuffer.kt
package tech.kzen.auto.server.objects.report.pipeline.input.model

import tech.kzen.auto.server.objects.report.pipeline.input.parse.CsvRecordParser
import tech.kzen.auto.server.objects.report.pipeline.input.parse.TsvRecordParser
import java.io.ByteArrayOutputStream
import java.io.OutputStreamWriter
import java.io.Writer


class RecordItemBuffer(
    expectedContentLength: Int = 0,
    expectedFieldCount: Int = 0
) {
    //-----------------------------------------------------------------------------------------------------------------
    companion object {
        fun toCsv(values: List<String>): String {
            return of(values).toCsv()
        }

        fun of(vararg values: String): RecordItemBuffer {
            return of(listOf(*values))
        }

        fun of(values: List<String>): RecordItemBuffer {
            val buffer = RecordItemBuffer()

            for (value in values) {
                for (i in value.indices) {
                    buffer.addToField(value[i])
                }
                buffer.commitField()
            }

            return buffer
        }

        fun of(contents: CharArray, offset: Int, length: Int): RecordItemBuffer {
            val buffer = RecordItemBuffer(length, 1)
            contents.copyInto(buffer.fieldContents, 0, offset, offset + length)
            buffer.fieldCount = 1
            buffer.fieldEnds[0] = length
            buffer.fieldContentLength = length
            buffer.nonEmpty = true
            return buffer
        }
    }


    //-----------------------------------------------------------------------------------------------------------------
    internal var fieldContents = CharArray(expectedContentLength)
    private var fieldEnds = IntArray(expectedFieldCount)
    private var fieldCount = 0
    private var fieldContentLength = 0
    private var nonEmpty = false

    val flyweight = RecordTextFlyweight(this)


    //-----------------------------------------------------------------------------------------------------------------
    fun selectFlyweight(fieldIndex: Int) {
        val startIndex: Int
        val endIndex: Int
        if (fieldIndex == 0) {
            startIndex = 0
            endIndex = fieldEnds[0]
        }
        else {
            startIndex = fieldEnds[fieldIndex - 1]
            endIndex = fieldEnds[fieldIndex]
        }

        val length = endIndex - startIndex

        flyweight.select(fieldIndex, startIndex, length)
    }


    fun size(): Int {
        return fieldCount
    }


    fun isEmpty(): Boolean {
        return ! nonEmpty && fieldCount <= 1 && fieldContentLength == 0
    }


    fun toList(): List<String> {
        val list = mutableListOf<String>()
        for (i in 0 until fieldCount) {
            val item = getString(i)
            list.add(item)
        }
        return list
    }


    //-----------------------------------------------------------------------------------------------------------------
    fun toCsv(): String {
        val out = ByteArrayOutputStream()
        OutputStreamWriter(out, Charsets.UTF_8).use {
            writeCsv(it)
        }
        return String(out.toByteArray(), Charsets.UTF_8)
    }


    fun writeCsv(out: Writer) {
        if (fieldCount == 1 && fieldContentLength == 0 && nonEmpty) {
            out.write("\"\"")
            return
        }

        for (i in 0 until fieldCount) {
            if (i != 0) {
                out.write(CsvRecordParser.delimiter)
            }

            writeCsvField(i, out)
        }
    }


    fun writeCsvField(index: Int, out: Writer) {
        val startIndex: Int
        val endIndex: Int
        if (index == 0) {
            startIndex = 0
            endIndex = fieldEnds[0]
        }
        else {
            startIndex = fieldEnds[index - 1]
            endIndex = fieldEnds[index]
        }

        CsvRecordParser.writeCsv(fieldContents, startIndex, endIndex, out);
    }


    fun toTsv(): String {
        val out = ByteArrayOutputStream()
        OutputStreamWriter(out, Charsets.UTF_8).use {
            writeTsv(it)
        }
        return String(out.toByteArray(), Charsets.UTF_8)
    }


    private fun writeTsv(out: Writer) {
        if (fieldCount == 1 && fieldContentLength == 0 && nonEmpty) {
            throw IllegalStateException("Can't represent non-empty record with single empty column")
        }

        for (i in 0 until fieldCount) {
            if (i != 0) {
                out.write(TsvRecordParser.delimiterInt)
            }

            writeTsvField(i, out)
        }
    }


    private fun writeTsvField(index: Int, out: Writer) {
        val startIndex: Int
        val endIndex: Int
        if (index == 0) {
            startIndex = 0
            endIndex = fieldEnds[0]
        }
        else {
            startIndex = fieldEnds[index - 1]
            endIndex = fieldEnds[index]
        }
        val length = endIndex - startIndex
        out.write(fieldContents, startIndex, length)
    }


    fun getString(index: Int): String {
        val startIndex: Int
        val endIndex: Int
        if (index == 0) {
            startIndex = 0
            endIndex = fieldEnds[0]
        }
        else {
            startIndex = fieldEnds[index - 1]
            endIndex = fieldEnds[index]
        }
        val length = endIndex - startIndex

        return String(fieldContents, startIndex, length)
    }


    //-----------------------------------------------------------------------------------------------------------------
    fun indicateNonEmpty() {
        nonEmpty = true
    }


    fun addToField(nextChar: Char) {
        growFieldContentsIfRequired()
        fieldContents[fieldContentLength] = nextChar
        fieldContentLength++
    }


    fun addToField(chars: CharArray, offset: Int, length: Int) {
        if (length != 0) {
            growFieldContentsIfRequired(fieldContentLength + length)
            System.arraycopy(chars, offset, fieldContents, fieldContentLength, length)
            fieldContentLength += length
        }
    }


    fun commitField() {
        growFieldEndsIfRequired()

        fieldEnds[fieldCount] = fieldContentLength
        fieldCount++
    }


    fun addToFieldAndCommit(chars: CharArray, offset: Int, length: Int) {
        if (length != 0) {
            if (fieldContents.size < fieldContentLength + length) {
                val nextSize = (fieldContents.size * 1.2).toInt().coerceAtLeast(fieldContentLength + length)
                fieldContents = fieldContents.copyOf(nextSize)
            }
            System.arraycopy(chars, offset, fieldContents, fieldContentLength, length)
            fieldContentLength += length
        }

        if (fieldEnds.size <= fieldCount) {
            val nextSize = (fieldEnds.size * 1.2 + 1).toInt().coerceAtLeast(fieldCount)
            fieldEnds = fieldEnds.copyOf(nextSize)
        }
        fieldEnds[fieldCount] = fieldContentLength
        fieldCount++
    }


    fun add(value: String) {
        for (i in value.indices) {
            addToField(value[i])
        }
        commitField()
    }


    fun addAll(values: List<String>) {
        for (value in values) {
            add(value)
        }
    }


    fun clear() {
        fieldCount = 0
        fieldContentLength = 0
        nonEmpty = false
    }


    //-----------------------------------------------------------------------------------------------------------------
    fun addToFieldUnsafe(nextChar: Char) {
        fieldContents[fieldContentLength] = nextChar
        fieldContentLength++
    }


    fun addToFieldUnsafe(chars: CharArray, offset: Int, length: Int) {
        System.arraycopy(chars, offset, fieldContents, fieldContentLength, length)
        fieldContentLength += length
    }


    fun commitFieldUnsafe() {
        fieldEnds[fieldCount] = fieldContentLength
        fieldCount++
    }


    fun addToFieldAndCommitUnsafe(chars: CharArray, offset: Int, length: Int) {
        System.arraycopy(chars, offset, fieldContents, fieldContentLength, length)
        fieldContentLength += length

        fieldEnds[fieldCount] = fieldContentLength
        fieldCount++
    }


    //-----------------------------------------------------------------------------------------------------------------
    fun copy(that: RecordItemBuffer) {
        fieldCount = that.fieldCount
        fieldContentLength = that.fieldContentLength
        nonEmpty = that.nonEmpty

        growFieldContentsIfRequired(fieldContentLength)
        growFieldEndsIfRequired()

        that.fieldContents.copyInto(fieldContents, endIndex = fieldContentLength)
        that.fieldEnds.copyInto(fieldEnds, endIndex = fieldCount)
    }


    fun prototype(): RecordItemBuffer {
        val prototype = RecordItemBuffer()
        prototype.copy(this)
        return prototype
    }


    //-----------------------------------------------------------------------------------------------------------------
    fun growTo(requiredLength: Int, requiredFieldCount: Int) {
        growFieldContentsIfRequired(requiredLength)
        growFieldEndsIfRequired(requiredFieldCount)
    }


    fun growBy(additionalLength: Int, additionalFieldCount: Int) {
        growFieldContentsIfRequired(fieldContentLength + additionalLength)
        growFieldEndsIfRequired(fieldCount + additionalFieldCount)
    }


    private fun growFieldContentsIfRequired(required: Int = fieldContentLength + 1) {
        if (fieldContents.size < required) {
            val nextSize = (fieldContents.size * 1.2).toInt().coerceAtLeast(required)
            fieldContents = fieldContents.copyOf(nextSize)
        }
    }


    private fun growFieldEndsIfRequired(required: Int = fieldCount) {
        if (fieldEnds.size <= required) {
            val nextSize = (fieldEnds.size * 1.2 + 1).toInt().coerceAtLeast(required)
            fieldEnds = fieldEnds.copyOf(nextSize)
        }
    }


    //-----------------------------------------------------------------------------------------------------------------
    override fun toString(): String {
        return toCsv()
    }
}